src/main/scala/org/ergoplatform/settings/Parameters.scala

```diff
@@ -1,30 +1,283 @@
+                   epochVotes: Seq[(Byte, Int)],
+                   votingEpochLength: Int): Map[Byte, Int] = {
+    epochVotes.filter(_._1 < Parameters.SoftFork).foldLeft(parametersTable) { case (table, (paramId, count)) =>
+
+      val paramIdAbs = if (paramId < 0) (-paramId).toByte else paramId
+
+      if (count > votingEpochLength / 2) {
+        val currentValue = parametersTable(paramIdAbs)
+        val maxValue = maxValues.getOrElse(paramIdAbs, Int.MaxValue / 2) //todo: more precise upper-bound
+        val minValue = minValues.getOrElse(paramIdAbs, 0)
+        val step = stepsTable.getOrElse(paramIdAbs, Math.max(1, currentValue / 100))
+
+        val newValue = paramId match {
+          case b: Byte if b > 0 =>
+            if (currentValue < maxValue) currentValue + step else currentValue
+          case b: Byte if b < 0 =>
+            if (currentValue > minValue) currentValue - step else currentValue
+        }
+        table.updated(paramIdAbs, newValue)
+      } else {
+        table
+      }
+    }
+  }
+
+  private def padVotes(vs: Array[Byte]): Array[Byte] = {
+    val maxVotes = ParamVotesCount + 1
+    if (vs.length < maxVotes) vs ++ Array.fill(maxVotes - vs.length)(0: Byte) else vs
+  }
+
+  def vote(ownTargets: Map[Byte, Int], epochVotes: Array[(Byte, Int)], voteForFork: Boolean): Array[Byte] = {
+    val vs = epochVotes.filter { case (paramId, _) =>
+      if(paramId == Parameters.SoftFork) {
+        voteForFork
+      } else if (paramId > 0) {
+        ownTargets.get(paramId).exists(_ > parametersTable(paramId))
+      } else if (paramId < 0) {
+        ownTargets.get((-paramId).toByte).exists(_ < parametersTable((-paramId).toByte))
+      } else {
+        false
+      }
+    }.map(_._1)
+    padVotes(vs)
+  }
+
+  def suggestVotes(ownTargets: Map[Byte, Int], voteForFork: Boolean): Array[Byte] = {
+    val vs = ownTargets.flatMap { case (paramId, value) =>
+      if (paramId == SoftFork) None
+      else if (value > parametersTable(paramId)) Some(paramId)
+      else if (value < parametersTable(paramId)) Some((-paramId).toByte)
+      else None
+    }.take(ParamVotesCount).toArray
+    padVotes(if (voteForFork) vs :+ SoftFork else vs)
+  }
+
+  def toExtensionCandidate(optionalFields: Seq[(Array[Byte], Array[Byte])] = Seq()): ExtensionCandidate = {
+    val mandatoryFields = parametersTable.toSeq.map { case (k, v) => Array(k) -> Ints.toByteArray(v) }
+    ExtensionCandidate(mandatoryFields, optionalFields)
+  }
+
+  override def toString: String = s"Parameters(height: $height; ${parametersTable.mkString("; ")})"
+}
+
+object Parameters {
+
+  val SoftFork = 120: Byte
+  val SoftForkVotesCollected = 121: Byte
+  val SoftForkStartingHeight = 122: Byte
+  val BlockVersion = 123: Byte
+
+  //A vote for nothing
+  val NoParameter = 0: Byte
+
+  //Parameter identifiers
+  val StorageFeeFactorIncrease = 1: Byte
+  val StorageFeeFactorDecrease = (-StorageFeeFactorIncrease).toByte
```

Comment:
Some `*Decrease` methods are unused, the rest are only used in tests. I propose to remain only one parameter identifier for all. 
Also public members should have type annotation.

Reply:
I would like to leave them for documentation purposes.